src/lib.rs

```diff
@@ -379,28 +378,23 @@ impl Rav1dPicture {
     }
 }
 
-#[allow(unsafe_op_in_unsafe_fn)]
-unsafe fn output_image(
-    c: &Rav1dContext,
-    state: &mut Rav1dState,
-    out: &mut Rav1dPicture,
-) -> Rav1dResult {
+fn output_image(c: &Rav1dContext, state: &mut Rav1dState, out: &mut Rav1dPicture) -> Rav1dResult {
     let mut res = Ok(());
 
-    let r#in: *mut Rav1dThreadPicture = if c.all_layers || state.max_spatial_id == 0 {
+    let r#in = if c.all_layers || state.max_spatial_id == 0 {
         &mut state.out
     } else {
         &mut state.cache
     };
-    if !c.apply_grain || !(*r#in).p.has_grain() {
-        *out = mem::take(&mut (*r#in).p);
+    if !c.apply_grain || !r#in.p.has_grain() {
+        *out = mem::take(&mut r#in.p);
     } else {
-        res = rav1d_apply_grain(c, out, &(*r#in).p);
+        res = rav1d_apply_grain(c, out, &r#in.p);
     }
-    let _ = mem::take(&mut *r#in);
+    let _ = mem::take(r#in);
 
     if !c.all_layers && state.max_spatial_id != 0 && state.out.p.data.is_some() {
```

Comment:
Nit: Could you pull out this expr, `!c.all_layers && state.max_spatial_id != 0` into a variable like `use_cache` and then also use `use_cache` where we initialize `r#in` above?